streamline drake systems interfaces inside drake_iiwa_sim

pangtao22 opened the pull request:
The controller method which calculates torque input to the robot was buried under several layers of drake system inheritance. This PR 

- creates a custom controller system that exposes torque calculation
- allows the controller to take reference torque as an input (which is harder to do in drake's implementation)
- constructs `Diagram` without using `SimDiagramBuilder`.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/spartan/309)
<!-- Reviewable:end -->

pangtao22 commented:
+@gizatt 

---

Review status: 0 of 4 files reviewed, all discussions resolved

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/spartan/309#-:-LFxzPS7CkeupoGZ13dy:b-mp15vn)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Reviewed 3 of 4 files at r1.
Review status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @pangtao22 and @gizatt)

---

*[src/catkin_projects/drake_iiwa_sim/src/kuka_simulation.cc, line 89 at r1](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFy1ANs2KM1hdmHCy9S:-LFy1ANs2KM1hdmHCy9T:b5jxe6g) ([raw file](https://github.com/robotlocomotion/spartan/blob/e95d3f9eec5799d2687e9c39373b63269d13ed8b/src/catkin_projects/drake_iiwa_sim/src/kuka_simulation.cc#L89)):*
> ```cpp
>         urdf, multibody::joints::kFixed, tree.get());
>     multibody::AddFlatTerrainToWorld(tree.get(), 100., 10.);
>     plant = builder.template AddSystem<systems::RigidBodyPlant<double>>(
> ```

Should there be a space between `template` and `AddSystem`?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/spartan/309)*
<!-- Sent from Reviewable.io -->

gizatt commented:
First pass complete, looks great! Have you been able to run this sim + control it from e.g. director?

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed, 5 unresolved discussions (waiting on @pangtao22)

---

*[src/catkin_projects/drake_iiwa_sim/include/drake_iiwa_sim/iiwa_qp_inverse_dynamics_controller.h, line 28 at r1](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFy-lEG2iwSGrEDLy7Z:-LFy-lEG2iwSGrEDLy7_:b-n3p8wl) ([raw file](https://github.com/robotlocomotion/spartan/blob/e95d3f9eec5799d2687e9c39373b63269d13ed8b/src/catkin_projects/drake_iiwa_sim/include/drake_iiwa_sim/iiwa_qp_inverse_dynamics_controller.h#L28)):*
> ```objc
> 
>   const systems::InputPortDescriptor<double>& get_input_port_estimated_state() const {
>     return this->get_input_port(0);
> ```

See below -- if you capture the InputPortDescriptor and OutputPortDescriptor returns from your DeclareInputPort calls, you could just return const references to those here and not have to worry about keeping these indices up to date.

---

*[src/catkin_projects/drake_iiwa_sim/src/CMakeLists.txt, line 44 at r1](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFy0KJK0pkI2Ej60joL:-LFy0KJK0pkI2Ej60joM:bqsp5u5) ([raw file](https://github.com/robotlocomotion/spartan/blob/e95d3f9eec5799d2687e9c39373b63269d13ed8b/src/catkin_projects/drake_iiwa_sim/src/CMakeLists.txt#L44)):*
> ```CMake
> add_executable(iiwa_sim kuka_simulation.cc)
> target_link_libraries(iiwa_sim 
> 	iiwa_lcm 
> ```

FYI prefer spaces, though nbd

---

*[src/catkin_projects/drake_iiwa_sim/src/iiwa_qp_inverse_dynamics_controller.cc, line 22 at r1](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFy0f612sI_42ekox2b:-LFy0f612sI_42ekox2c:b-lgfac3) ([raw file](https://github.com/robotlocomotion/spartan/blob/e95d3f9eec5799d2687e9c39373b63269d13ed8b/src/catkin_projects/drake_iiwa_sim/src/iiwa_qp_inverse_dynamics_controller.cc#L22)):*
> ```cpp
>     : control_period_(control_period) {
>   // copy stuff.
>   tree_ = std::move(tree);
> ```

This being a unique ptr seems sort of weird, since that tree probably gets passed around to lots of different places... is this how it has been done in other places in Drake / Drake-ported code?

---

*[src/catkin_projects/drake_iiwa_sim/src/iiwa_qp_inverse_dynamics_controller.cc, line 23 at r1](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFy0mXBCnh31ooglFl6:-LFy0mXBCnh31ooglFl7:bj3thrb) ([raw file](https://github.com/robotlocomotion/spartan/blob/e95d3f9eec5799d2687e9c39373b63269d13ed8b/src/catkin_projects/drake_iiwa_sim/src/iiwa_qp_inverse_dynamics_controller.cc#L23)):*
> ```cpp
>   // copy stuff.
>   tree_ = std::move(tree);
>   kp_ = kp;
> ```

Maybe sanity-check the sizes of Kp and Kd, especially since their size is referenced when creating input ports later on? And while you're at it, sanity checking that nq_ matches the actual size of the rbt might be good to detect important changes / mistakes in the construction of the IIWA RBT.

---

*[src/catkin_projects/drake_iiwa_sim/src/iiwa_qp_inverse_dynamics_controller.cc, line 29 at r1](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFy-MzJAl0vc2ryrAP6:-LFy-MzKEj6FP9L5IhPK:b-hhky10) ([raw file](https://github.com/robotlocomotion/spartan/blob/e95d3f9eec5799d2687e9c39373b63269d13ed8b/src/catkin_projects/drake_iiwa_sim/src/iiwa_qp_inverse_dynamics_controller.cc#L29)):*
> ```cpp
>   this->DeclarePeriodicDiscreteUpdate(control_period_);
>   // estimated state input port
>   this->DeclareVectorInputPort(systems::BasicVector<double>(2 * nq_));
> ```

This function returns an InputPortDescriptor, which you could store as a private class member and use to find the appropriate input ports in those helper functions in your header. (Ditto for output port). Will be a lot more robust to future tweaks to this class' input and output port declaration ordering

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/spartan/309#-:-LFy1j42-WHJyPNDqEr3:by6ydkv)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Review status: all files reviewed, 5 unresolved discussions (waiting on @pangtao22)

---

*[src/catkin_projects/drake_iiwa_sim/src/CMakeLists.txt, line 44 at r1](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFy0KJK0pkI2Ej60joL:-LFy25TH1KE7FV6ieNVk:b-d8quyr) ([raw file](https://github.com/robotlocomotion/spartan/blob/e95d3f9eec5799d2687e9c39373b63269d13ed8b/src/catkin_projects/drake_iiwa_sim/src/CMakeLists.txt#L44)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

FYI prefer spaces, though nbd
</blockquote></details>

:+1: spaces not tabs

---

*[src/catkin_projects/drake_iiwa_sim/src/iiwa_qp_inverse_dynamics_controller.cc, line 29 at r1](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFy-MzJAl0vc2ryrAP6:-LFy2G760n40UUJWp4gT:bmchxvu) ([raw file](https://github.com/robotlocomotion/spartan/blob/e95d3f9eec5799d2687e9c39373b63269d13ed8b/src/catkin_projects/drake_iiwa_sim/src/iiwa_qp_inverse_dynamics_controller.cc#L29)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

This function returns an InputPortDescriptor, which you could store as a private class member and use to find the appropriate input ports in those helper functions in your header. (Ditto for output port). Will be a lot more robust to future tweaks to this class' input and output port declaration ordering
</blockquote></details>

Agreed. I am not as up on drake system stuff as Greg, but any way that we can avoid having random indices flying around that have to be matched up would be great.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/spartan/309)*
<!-- Sent from Reviewable.io -->

gizatt commented:
Review status: all files reviewed, 5 unresolved discussions (waiting on @pangtao22 and @gizatt)

---

*[src/catkin_projects/drake_iiwa_sim/src/kuka_simulation.cc, line 89 at r1](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFy1ANs2KM1hdmHCy9S:-LFy2NE-Cbw8IfUZ2w_b:br3q5n6) ([raw file](https://github.com/robotlocomotion/spartan/blob/e95d3f9eec5799d2687e9c39373b63269d13ed8b/src/catkin_projects/drake_iiwa_sim/src/kuka_simulation.cc#L89)):*
<details><summary><i>Previously, manuelli (Lucas Manuelli) wrote…</i></summary><blockquote>

Should there be a space between `template` and `AddSystem`?
</blockquote></details>

It's template black magic (necessary to disambiguate the use of the templated method of builder, I think) https://www.ibm.com/support/knowledgecenter/en/SSLTBW_2.3.0/com.ibm.zos.v2r3.cbclx01/keyword_template_qualifier.htm

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/spartan/309)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Review status: all files reviewed, 5 unresolved discussions (waiting on @pangtao22 and @gizatt)

---

*[src/catkin_projects/drake_iiwa_sim/src/kuka_simulation.cc, line 89 at r1](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFy1ANs2KM1hdmHCy9S:-LFy3R5F2XoayMH_nAmb:bgbmse) ([raw file](https://github.com/robotlocomotion/spartan/blob/e95d3f9eec5799d2687e9c39373b63269d13ed8b/src/catkin_projects/drake_iiwa_sim/src/kuka_simulation.cc#L89)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

It's template black magic (necessary to disambiguate the use of the templated method of builder, I think) https://www.ibm.com/support/knowledgecenter/en/SSLTBW_2.3.0/com.ibm.zos.v2r3.cbclx01/keyword_template_qualifier.htm
</blockquote></details>

crazy, i've never seen that. as long as it works I'm fine with it.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/spartan/309)*
<!-- Sent from Reviewable.io -->

pangtao22 commented:
Yup. And it works as expected. 

---

Review status: 0 of 4 files reviewed, 5 unresolved discussions (waiting on @gizatt, @manuelli, and @pangtao22)

---

*[src/catkin_projects/drake_iiwa_sim/include/drake_iiwa_sim/iiwa_qp_inverse_dynamics_controller.h, line 28 at r1](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFy-lEG2iwSGrEDLy7Z:-LFz58VU8xAfCen_tlfW:b-896fix) ([raw file](https://github.com/robotlocomotion/spartan/blob/e95d3f9eec5799d2687e9c39373b63269d13ed8b/src/catkin_projects/drake_iiwa_sim/include/drake_iiwa_sim/iiwa_qp_inverse_dynamics_controller.h#L28)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

See below -- if you capture the InputPortDescriptor and OutputPortDescriptor returns from your DeclareInputPort calls, you could just return const references to those here and not have to worry about keeping these indices up to date.
</blockquote></details>

Done.

---

*[src/catkin_projects/drake_iiwa_sim/src/CMakeLists.txt, line 44 at r1](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFy0KJK0pkI2Ej60joL:-LFz5NHk2w2PEsnagLZV:b-crzrlm) ([raw file](https://github.com/robotlocomotion/spartan/blob/e95d3f9eec5799d2687e9c39373b63269d13ed8b/src/catkin_projects/drake_iiwa_sim/src/CMakeLists.txt#L44)):*
<details><summary><i>Previously, manuelli (Lucas Manuelli) wrote…</i></summary><blockquote>

:+1: spaces not tabs
</blockquote></details>

Changed to spaces, I think (by setting tab as spaces in my editor).

---

*[src/catkin_projects/drake_iiwa_sim/src/iiwa_qp_inverse_dynamics_controller.cc, line 22 at r1](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFy0f612sI_42ekox2b:-LFy7rxf9RQDDYJ3LQup:b-ixvso0) ([raw file](https://github.com/robotlocomotion/spartan/blob/e95d3f9eec5799d2687e9c39373b63269d13ed8b/src/catkin_projects/drake_iiwa_sim/src/iiwa_qp_inverse_dynamics_controller.cc#L22)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

This being a unique ptr seems sort of weird, since that tree probably gets passed around to lots of different places... is this how it has been done in other places in Drake / Drake-ported code?
</blockquote></details>

In Drake's `iiwa_simulation`, every drake system using a `rigidbodytree` has its own copy of the tree. I am not sure why they did that. 

---

*[src/catkin_projects/drake_iiwa_sim/src/iiwa_qp_inverse_dynamics_controller.cc, line 23 at r1](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFy0mXBCnh31ooglFl6:-LFz7YXeDCD4Qf4obxfg:bxoq613) ([raw file](https://github.com/robotlocomotion/spartan/blob/e95d3f9eec5799d2687e9c39373b63269d13ed8b/src/catkin_projects/drake_iiwa_sim/src/iiwa_qp_inverse_dynamics_controller.cc#L23)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

Maybe sanity-check the sizes of Kp and Kd, especially since their size is referenced when creating input ports later on? And while you're at it, sanity checking that nq_ matches the actual size of the rbt might be good to detect important changes / mistakes in the construction of the IIWA RBT.
</blockquote></details>

What's the best way to check? I added a few assertions, but they are ignored when compiled in release mode. 

---

*[src/catkin_projects/drake_iiwa_sim/src/iiwa_qp_inverse_dynamics_controller.cc, line 29 at r1](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFy-MzJAl0vc2ryrAP6:-LFz5tvB29y-EDtDRvLD:b-896fix) ([raw file](https://github.com/robotlocomotion/spartan/blob/e95d3f9eec5799d2687e9c39373b63269d13ed8b/src/catkin_projects/drake_iiwa_sim/src/iiwa_qp_inverse_dynamics_controller.cc#L29)):*
<details><summary><i>Previously, manuelli (Lucas Manuelli) wrote…</i></summary><blockquote>

Agreed. I am not as up on drake system stuff as Greg, but any way that we can avoid having random indices flying around that have to be matched up would be great.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/spartan/309#-:-LFy9mLmESq7l1MP-c1f:byfmn2n)*
<!-- Sent from Reviewable.io -->

pangtao22 commented:
Review status: 0 of 4 files reviewed, 5 unresolved discussions (waiting on @gizatt and @manuelli)

---

*[src/catkin_projects/drake_iiwa_sim/src/CMakeLists.txt, line 33 at r2](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFzCO060l8zCKwVp0zR:-LFzCO060l8zCKwVp0zS:brb0eag) ([raw file](https://github.com/robotlocomotion/spartan/blob/2321426f7099a608038a2fd7994cb9fa9e1816b9/src/catkin_projects/drake_iiwa_sim/src/CMakeLists.txt#L33)):*
> ```CMake
> # POSSIBILITY OF SUCH DAMAGE.
> 
> set(PROJECT_INCLUDE_DIR ${PROJECT_SOURCE_DIR}/include/drake_iiwa_sim/)
> ```

I figured out a way to use Clion inside docker. These lines are added so that CLion will do static analysis on the header files. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/spartan/309)*
<!-- Sent from Reviewable.io -->

gizatt commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, a few tweaks that you could make if you want but merge at your discretion. (Lucas technically has a blocking comment but I think that chain has been resolved.)

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed, 1 unresolved discussion (waiting on @manuelli)

---

*[src/catkin_projects/drake_iiwa_sim/include/drake_iiwa_sim/iiwa_qp_inverse_dynamics_controller.h, line 28 at r1](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFy-lEG2iwSGrEDLy7Z:-LG1grAp8weAm2VhMYJg:b-1gbaei) ([raw file](https://github.com/robotlocomotion/spartan/blob/e95d3f9eec5799d2687e9c39373b63269d13ed8b/src/catkin_projects/drake_iiwa_sim/include/drake_iiwa_sim/iiwa_qp_inverse_dynamics_controller.h#L28)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

Done.
</blockquote></details>

FYI you could store the input and output port descriptor references themselves, I think, and skip the extra step of storing / accessing them by index. But this works fine!

---

*[src/catkin_projects/drake_iiwa_sim/src/CMakeLists.txt, line 33 at r2](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFzCO060l8zCKwVp0zR:-LG1h4hq8XChyXhrRmj-:bnqu6xc) ([raw file](https://github.com/robotlocomotion/spartan/blob/2321426f7099a608038a2fd7994cb9fa9e1816b9/src/catkin_projects/drake_iiwa_sim/src/CMakeLists.txt#L33)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

I figured out a way to use Clion inside docker. These lines are added so that CLion will do static analysis on the header files. 
</blockquote></details>

Neat! Not a Clion user myself (I don't think most of the Spartan devs are?) but I'm OK with this!

---

*[src/catkin_projects/drake_iiwa_sim/src/iiwa_qp_inverse_dynamics_controller.cc, line 22 at r1](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFy0f612sI_42ekox2b:-LG1hDuIDAQWQnygQBII:b-a12eo8) ([raw file](https://github.com/robotlocomotion/spartan/blob/e95d3f9eec5799d2687e9c39373b63269d13ed8b/src/catkin_projects/drake_iiwa_sim/src/iiwa_qp_inverse_dynamics_controller.cc#L22)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

In Drake's `iiwa_simulation`, every drake system using a `rigidbodytree` has its own copy of the tree. I am not sure why they did that. 
</blockquote></details>

Huh, ok! I have a vague feeling this might have been a passing phase in TRI's design that has since gone away, but I've been in Python land for too long and could be wrong. Fine either way

---

*[src/catkin_projects/drake_iiwa_sim/src/iiwa_qp_inverse_dynamics_controller.cc, line 23 at r1](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFy0mXBCnh31ooglFl6:-LG1hhPw8mOE6TAqXJyV:b-nd7x9w) ([raw file](https://github.com/robotlocomotion/spartan/blob/e95d3f9eec5799d2687e9c39373b63269d13ed8b/src/catkin_projects/drake_iiwa_sim/src/iiwa_qp_inverse_dynamics_controller.cc#L23)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

What's the best way to check? I added a few assertions, but they are ignored when compiled in release mode. 
</blockquote></details>

Debug-only assertions seem OK, as when we're developing code, we should probably be in debug mode :). But I think more generic C++ asserts() might still work in release mode -- or failing that, I'm down for some if statements + prints to warn the user.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/spartan/309#-:-LG1i4r-56_6XROu9l0u:b-7gyx1z)*
<!-- Sent from Reviewable.io -->

pangtao22 commented:
Dismissed @manuelli from a discussion.
Review status: :shipit: complete! all files reviewed, all discussions resolved

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/spartan/309)*
<!-- Sent from Reviewable.io -->

pangtao22 commented:
Review status: :shipit: complete! all files reviewed, all discussions resolved

---

*[src/catkin_projects/drake_iiwa_sim/src/CMakeLists.txt, line 33 at r2](https://reviewable.io/reviews/robotlocomotion/spartan/309#-LFzCO060l8zCKwVp0zR:-LG1oQ3_DTTtQzd6avkV:b-ue2olo) ([raw file](https://github.com/robotlocomotion/spartan/blob/2321426f7099a608038a2fd7994cb9fa9e1816b9/src/catkin_projects/drake_iiwa_sim/src/CMakeLists.txt#L33)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

Neat! Not a Clion user myself (I don't think most of the Spartan devs are?) but I'm OK with this!
</blockquote></details>

I really like being able to see almost any identifier's type just by hovering the mouse on it. I'll add a section to the README on how to setup clion. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/spartan/309)*
<!-- Sent from Reviewable.io -->
